### PR TITLE
Rust connection pool/Python integration

### PR DESCRIPTION
**Pull request by mmastrac**
Follow-up work from https://github.com/edgedb/edgedb/pull/7478

Note that running `EDGEDB_USE_NEW_CONNPOOL=1 edb test` fails with a few tests. These will get sorted out as we move forward. The tests unfortunately require the logging to match the expected output from the old pool, which is a small challenge. I'll figure out if it makes sense to rewrite these tests in the next round.

The interaction between Rust and Python is pretty straightforward: we use a cross-thread channel to send messages and we pair that with a self-pipe that we write to to notify the asyncio end that there's a new message. We preemptively read in a few places to reduce latency, and if we manage to pick up a message, reduce the number of messages we expect to read async by one.

As we move forward, these patterns will probably get moved into a more common set of code rather than being adhoc like this.

What's left after this PR:

 - Complete garbage collection work -- connections should not get GC'd too quickly
 - Export metrics/logs

@msullivan I'd recommend looking at the following files:

 - `edb/server/conn_pool/src/python.rs` - the Rust/Python integration strategy
 - `edb/server/connpool/` - the actual Python end of things (and changes required to ensure we can run two connpools side-by-side)

@fantix It would be great if you could take a look at the Rust changes. For the most part they are just refactorings, a small bugfix in the drain code and some light algorithm tweaks.

### CODE REVIEW

File: edb/server/connpool/pool2.py [hunk 1]
```diff
@@ -1,5 +1,361 @@
...
+    failed_connects: int
+    failed_disconnects: int
+    successful_connects: int
+    successful_disconnects: int
+
+
+class StatsCollector(typing.Protocol):
+
+    def __call__(self, stats: Snapshot) -> None:
+        pass
```
[the remark this answers is not in the document]
Comment (mmastrac) [reply edb/server/connpool/pool2.py:84]: I think so -- this is the same type as lives in pool.py. I can try to fix this while I'm here.

File: edb/server/conn_pool/src/python.rs [hunk 2]
```diff
@@ -1,6 +1,468 @@
...
+    }
+
+    fn _prune(&self, id: u64, db: &str) -> PyResult<()> {
+        self.python_to_rust
+            .send(PythonToRustMessage::Prune(id, db.to_owned()))
+            .map_err(|_| internal_error("In shutdown"))
+    }
+
+    fn _read(&self, py: Python<'_>) -> Py<PyAny> {
+        let Ok(msg) = self.rust_to_python.recv() else {
```
[the remark this answers is not in the document]
Comment (mmastrac) [reply edb/server/conn_pool/src/python.rs:354]: In this case we can be 100% confident that there's something in the pipe, so this recv() will never block and we'll either get a result, or we'll get the "end of data" from the channel. I can add some extra comments around this.

File: edb/server/connpool/pool2.py [hunk 3]
```diff
@@ -1,5 +1,361 @@
...
+            error = self._errors.pop(msg[2], None)
+            if error:
+                if f := self._acquires.pop(msg[1], None):
+                    f.set_exception(error)
+                else:
+                    logger.warn(f"Duplicate exception for acquire {msg[1]}")
+        else:
+            logger.critical(f'Unexpected message: {msg}')
+
+    async def _perform_connect(self, id: int, db: str) -> None:
```
[the remark this answers is not in the document]
Comment (mmastrac) [reply edb/server/connpool/pool2.py:217]: Thanks for the tip, I'll change this!

File: edb/server/conn_pool/src/pool.rs [hunk 4]
```diff
@@ -131,12 +127,20 @@ impl<C: Connector> Pool<C> {
             blocks: Default::default(),
             connector,
             dirty: Default::default(),
-            drain: Default::default(),
+            drain: Drain::default(),
```
[the remark this answers is not in the document]
Comment (mmastrac) [reply edb/server/conn_pool/src/pool.rs:130]: Unfortunately that only works for types where all fields are `: Default`. 😢 There are some libraries that allow for partial defaults.

File: edb/server/connpool/pool2.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (fantix) [reply edb/server/connpool/pool2.py:84]: Oh, I see - it's okay either way.

File: edb/testbase/server.py [hunk 5]
```diff
@@ -2451,6 +2451,32 @@ async def __aenter__(self):
                 timeout=240,
             ),
         )
+
+        output = b''
+
+        async def read_stdout():
```
[the remark this answers is not in the document]
Comment (mmastrac) [reply edb/testbase/server.py]: Discussed offline and we'll punt this until a later time.

File: edb/server/connpool/pool2.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mmastrac) [reply edb/server/connpool/pool2.py:216]: Discussed offline and we'll skip this, but noted for the future

**Pull request merged**